pkg/status/singletonstatus.go

```diff
@@ -0,0 +1,233 @@
+/*
+Copyright 2024 The KubeStellar Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package status
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"github.com/kubestellar/kubestellar/api/control/v1alpha1"
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+// reconcileSingletonByBdg goes over the binding-covered workload objects, decides sync or not for each of the workload objects,
+// then maintains their statuses based on the decisions.
+func (c *Controller) reconcileSingletonByBdg(ctx context.Context, bdgName string) error {
+	logger := klog.FromContext(ctx)
+	logger.V(4).Info("Reconciling singleton status due to binding change", "binding", bdgName)
+
+	binding, err := c.wdsKsClient.ControlV1alpha1().Bindings().Get(ctx, bdgName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("could not get binding %s from cache", bdgName)
+	}
+
+	sync := make(map[util.ObjectIdentifier]bool)
+	if err = c.checkSingletonWorkloadInBinding(ctx, *binding, sync, true); err != nil {
+		return err
+	}
+
+	allBdgs, err := c.wdsKsClient.ControlV1alpha1().Bindings().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, bdg := range allBdgs.Items {
+		if bdg.Name == binding.Name {
+			continue
+		}
+		if err = c.checkSingletonWorkloadInBinding(ctx, bdg, sync, false); err != nil {
+			return err
+		}
+	}
+
+	for wObjIdentifier, v := range sync {
+		if err := c.reconcileSingletonWObj(ctx, wObjIdentifier, v); err != nil {
+			return err
+		}
+	}
```

Comment:
What this logic does is: first determine whether any Binding that is not being deleted associates an object in `sync` with at least one WEC, and then for each such object copy the status from *every* related WorkStatus to the status section of that object in the WDS. **Singleton** status return is supposed to be about returning status for an object that is associated with _exactly one_ WEC.